src/methods.jl

```diff
@@ -258,63 +335,134 @@ estimate of the derivative.
-    factor::Real=1,
-    max_step::Real=0.1 * max(abs(x), one(x))
-) where T<:AbstractFloat
-    p = length(m.coefs)
-    q = m.q
+    m::UnadaptedFiniteDifferenceMethod, f::TF, x::T,
+) where {TF<:Function,T<:AbstractFloat}
+    step, acc = _compute_step_acc_default(m, x)
+    return _limit_step(m, x, step, acc)
+end
+function estimate_step(
+    m::AdaptedFiniteDifferenceMethod{P,Q}, f::TF, x::T,
+) where {P,Q,TF<:Function,T<:AbstractFloat}
+    ∇f_magnitude, f_magnitude = _estimate_magnitudes(m.bound_estimator, f, x)
+    if ∇f_magnitude == 0.0 || f_magnitude == 0.0
+        step, acc = _compute_step_acc_default(m, x)
+    else
+        step, acc = _compute_step_acc(m, ∇f_magnitude, eps(f_magnitude))
+    end
+    return _limit_step(m, x, step, acc)
+end
 
-    # Estimate the round-off error.
-    ε = estimate_roundoff_error(f, x) * factor
+function _estimate_magnitudes(
+    m::FiniteDifferenceMethod{P,Q}, f::TF, x::T,
+) where {P,Q,TF<:Function,T<:AbstractFloat}
+    step = first(estimate_step(m, f, x))
+    fs = _eval_function(m, f, x, step)
+    # Estimate magnitude of `∇f` in a neighbourhood of `x`.
+    ∇fs = SVector{3}(
+        _compute_estimate(m, fs, x, step, m.coefs_neighbourhood[1]),
+        _compute_estimate(m, fs, x, step, m.coefs_neighbourhood[2]),
+        _compute_estimate(m, fs, x, step, m.coefs_neighbourhood[3])
+    )
+    ∇f_magnitude = maximum(maximum.(abs, ∇fs))
+    # Estimate magnitude of `f` in a neighbourhood of `x`.
+    f_magnitude = maximum(maximum.(abs, fs))
+    return ∇f_magnitude, f_magnitude
+end
 
-    # Estimate the bound on the derivatives.
-    M = m.bound_estimator(f, x)
+function _compute_step_acc_default(m::FiniteDifferenceMethod, x::T) where {T<:AbstractFloat}
+    # Compute a default step size using a heuristic and [`DEFAULT_CONDITION`](@ref).
+    return _compute_step_acc(m, m.condition, eps(T))
+end
 
+function _compute_step_acc(
+    m::FiniteDifferenceMethod{P,Q}, ∇f_magnitude::Real, f_error::Real,
+) where {P,Q}
     # Set the step size by minimising an upper bound on the error of the estimate.
-    C₁ = ε * sum(abs, m.coefs)
-    C₂ = M * sum(n -> abs(m.coefs[n] * m.grid[n]^p), eachindex(m.coefs)) / factorial(p)
-    # Type inference fails on this, so we annotate it, which gives big performance benefits.
-    h::T = convert(T, min((q / (p - q) * (C₁ / C₂))^(1 / p), max_step))
-
+    C₁ = f_error * m.f_error_mult * m.factor
+    C₂ = ∇f_magnitude * m.∇f_magnitude_mult
+    step = (Q / (P - Q) * (C₁ / C₂))^(1 / P)
     # Estimate the accuracy of the method.
-    accuracy = h^(-q) * C₁ + h^(p - q) * C₂
+    acc = C₁ * step^(-Q) + C₂ * step^(P - Q)
+    return step, acc
+end
 
-    return h, accuracy
+function _limit_step(
+    m::FiniteDifferenceMethod, x::T, step::Real, acc::Real,
+) where {T<:AbstractFloat}
+    # First, limit the step size based on the maximum range.
+    step_max = m.max_range / maximum(abs.(m.grid))
+    if step > step_max
+        step = step_max
+        acc = NaN
```

Comment:
is it documented somewhere what this means?

Also should we just stop computing accuracy?
Its never used nor exposed to the user AFAICT

Reply:
The function `estimate_step` is exposed to the user, which promises to also estimate the accuracy. You're right that `NaN` isn't documented, which I'll add.